File: public/socket.js
export const LOG_LEVEL_ERROR = 1;
export const LOG_LEVEL_DEBUG = 2;

// Idea based on URL: https://github.com/socketio/socket.io/blob/main/examples/basic-websocket-client/src/index.js
export class Socket {
    #subscribers = new Map();
    #emit(event, ...args) {
        if (!this.#subscribers.has(event)) {
            return;
        }

        const fns = this.#subscribers.get(event);
        for (const fn of fns) {
            fn(...args);
        }
    }
    #on(event, fn) {
        if (!this.#subscribers.has(event)) {
            this.#subscribers.set(event, [fn]);
            return;
        }

        const fns = this.#subscribers.get(event);
        fns.push(fn);
    }
    #off(event, fn) {
        if (event === undefined && fn === undefined) {
            this.#subscribers.clear();
            return;
        }

        if (!this.#subscribers.has(event)) {
            return;
        }

        if (fn === undefined) {
            this.#subscribers.set(event, []);
            return;
        }

        const fns = this.#subscribers.get(event);

        // Avoid "index shifting" issues
        for (let i = fns.length - 1; i >= 0; i--) {
            if (fns[i] === fn) {
                fns.splice(i, 1);
                break;
            }
        }
    }

    #url = undefined;
    #ws = undefined;

    #connected = false;
    #id = undefined;

    #ackId = 0;
    #ackFns = new Map();

    #level = LOG_LEVEL_DEBUG;

    constructor(url) {
        this.#url = url;
        this.connect();
    }

    #onPacket(packet) {
        switch (packet.type) {
            case 'connect':
                this.#onConnect(packet.data.id);
                break;
            case 'disconnect':
                this.#onDisconnect(packet.data.reason);
                break;
            case 'ack':
                if (this.#ackFns.has(packet.data.id)) {
                    const ackFn = this.#ackFns.get(packet.data.id);
                    ackFn(...packet.data.args);
                    this.#ackFns.delete(packet.data.id);
                }
                break;
            case 'event':
                if (packet.data.ackId > 0) {
                    packet.data.args.push((...args) => {
                        this.#emitAck(packet.data.ackId, ...args);
                    });
                }
                this.#emit(packet.data.event, ...packet.data.args);
                break;
            default:
                this.debug('Unknown packet type:', packet);
                break;
        }
    }

    #onConnect(id) {
        this.#connected = true;
        this.#id = id;

        this.#emit('connect', id);
    }

    connect(url) {
        if (this.#connected) {
            return this;
        }

        this.#ws = new WebSocket(this.#url);
        this.#ws.onopen = (evt) => {
            this.debug('ONOPEN HANDLER', evt);
        };
        this.#ws.onerror = (evt) => {
            this.debug('ONERROR HANDLER', evt);
        };
        this.#ws.onclose = (evt) => {
            this.debug('ONCLOSE HANDLER', evt);
            this.#onDisconnect('socket server disconnected');
        };
        this.#ws.onmessage = (evt) => {
            this.debug('ONMESSAGE HANDLER', evt);
            const packet = JSON.parse(evt.data);
            this.#onPacket(packet);
        };
        return this;
    }

    #onDisconnect(reason) {
        if (!this.#connected) {
            return;
        }

        this.#connected = false;
        this.#id = undefined;

        this.#ws.onclose;
        this.#ws.close();
        this.#ws = undefined;

        this.#ackId = 0;
        this.#ackFns.clear();

        this.#emit('disconnect', reason);
    }

    disconnect() {
        if (!this.#connected) {
            return;
        }

        this.#ws.close();
        return this;
    }

    get id() {
        return this.#id;
    }

    get connected() {
        return this.#connected;
    }

    get disconnected() {
        return !this.#connected;
    }

    emit(event, ...args) {
        if (!this.#connected) {
            return;
        }

        const ackFn = args.at(-1);
        const hasAckFn = typeof ackFn === 'function';
        if (hasAckFn) {
            this.#ackId += 1;
            this.#ackFns.set(this.#ackId, ackFn);

            // Remove the "ack" function
            args.pop();
        }

        const packet = JSON.stringify({
            type: 'event',
            data: {
                event,
                args: args,
                ackId: hasAckFn ? this.#ackId : 0,
            },
        });
        this.#ws.send(packet);
        return true;
    }

    #emitAck(id, ...args) {
        const packet = JSON.stringify({
            type: 'ack',
            data: {
                id: id,
                args: args,
            },
        });
        this.#ws.send(packet);
    }

    on(event, fn) {
        this.#on(event, fn);
        return this;
    }

    off(event, fn) {
        this.#off(event, fn);
        return this;
    }

    debug(...args) {
        if (this.#level <= LOG_LEVEL_DEBUG) {
            console.log(...args);
        }
    }

    setLogLevel(level) {
        this.#level = level;
    }
}

export function io(url) {
    const socket = new Socket(url);
    return socket;
}

File: pkg/socket/socket.go
package socket

import (
	"fmt"
	"log"
	"slices"

	"github.com/softwarespot/chatterbox/pkg/room"
)

type empty struct{}

type Packet struct {
	Type string         `json:"type"`
	Data map[string]any `json:"data"`
}

type Socket struct {
	subscribers map[string][]func(args ...any)

	adapter Adapter
	client  *room.Client[Args]

	connected bool

	ackID  int
	ackFns map[int]func(...any)

	disconnectedCh chan empty
}

func New(adapter Adapter) (*Socket, error) {
	s := &Socket{
		subscribers: map[string][]func(args ...any){},

		adapter: adapter,
		client:  nil,

		connected: false,

		ackID:  0,
		ackFns: map[int]func(...any){},

		disconnectedCh: make(chan empty),
	}

	var err error
	if s.client, err = room.NewClient[Args](); err != nil {
		return nil, err
	}

	go s.onPacket()

	return s, nil
}

func (s *Socket) emit(event string, args ...any) {
	for _, fn := range s.subscribers[event] {
		fn(args...)
	}
}

func (s *Socket) on(event string, fn func(args ...any)) {
	s.subscribers[event] = append(s.subscribers[event], fn)
}

func (s *Socket) off(event string, fn func(args ...any)) {
	if event == "" && fn == nil {
		clear(s.subscribers)
		return
	}

	fns, ok := s.subscribers[event]
	if !ok {
		return
	}

	if fn == nil {
		delete(s.subscribers, event)
		return
	}

	s.subscribers[event] = slices.DeleteFunc(fns, func(eventFn func(args ...any)) bool {
		// Compare the func references are the same
		return &fn == &eventFn
	})
}

func (s *Socket) onPacket() {
	for {
		var pkt Packet
		pkt, err := s.adapter.Receive()
		if err != nil {
			s.onDisconnect(err)
			break
		}

		switch pkt.Type {
		case "ack":
			id, ok := pkt.Data["id"].(float64)
			if !ok {
				log.Printf("invalid id type: %v", pkt.Data["id"])
				continue
			}
			ackID := int(id)

			args, ok := pkt.Data["args"].([]any)
			if !ok {
				log.Printf("invalid args type: %v", pkt.Data["args"])
				continue
			}
			if ackFn, ok := s.ackFns[ackID]; ok {
				ackFn(args...)
				delete(s.ackFns, ackID)
			}
		case "event":
			event, ok := pkt.Data["event"].(string)
			if !ok {
				log.Printf("invalid event type: %v", pkt.Data["event"])
				continue
			}

			args, ok := pkt.Data["args"].([]any)
			if !ok {
				log.Printf("invalid args type: %v", pkt.Data["args"])
				continue
			}

			id, ok := pkt.Data["ackId"].(float64)
			if !ok {
				log.Printf("invalid ackId type: %v", pkt.Data["ackId"])
				continue
			}
			ackID := int(id)

			if ackID > 0 {
				args = append(args, func(args ...any) {
					s.emitAck(ackID, args...)
				})
			}
			s.emit(event, args...)
		}
	}
}

func (s *Socket) onConnect() error {
	err := s.adapter.Send(Packet{
		Type: "connect",
		Data: map[string]any{
			"id": s.ID(),
		},
	})
	if err != nil {
		return fmt.Errorf("socket: sending connect packet: %w", err)
	}

	s.connected = true

	s.emit("connect", s.ID())

	<-s.disconnectedCh
	return nil
}

func (s *Socket) onDisconnect(err error) error {
	defer close(s.disconnectedCh)

	if !s.connected {
		return nil
	}

	reason := err.Error()
	err = s.adapter.Send(Packet{
		Type: "disconnect",
		Data: map[string]any{
			"reason": reason,
		},
	})
	if err != nil {
		return fmt.Errorf("socket: sending disconnect packet: %w", err)
	}

	s.connected = false
	s.client = nil

	s.ackID = 0
	clear(s.ackFns)

	s.emit("disconnect", reason)

	if err := s.adapter.Close(); err != nil {
		return fmt.Errorf("socket: closing websocket connection: %w", err)
	}
	return nil
}

func (s *Socket) ID() string {
	if s.client == nil {
		return ""
	}
	return s.client.ID()
}

func (s *Socket) Client() *room.Client[Args] {
	return s.client
}

func (s *Socket) Connected() bool {
	return s.connected
}

func (s *Socket) Disconnected() bool {
	return !s.connected
}

func (s *Socket) Emit(event string, args ...any) error {
	var ackID int
	if ackFn, ok := GetAckFunc(args); ok {
		s.ackID++
		s.ackFns[s.ackID] = ackFn

		// Remove the "ack" function
		args = argDeleteLast(args)
		ackID = s.ackID
	}

	err := s.adapter.Send(Packet{
		Type: "event",
		Data: map[string]any{
			"event": event,
			"args":  ensureNonEmptyArgs(args),
			"ackId": ackID,
		},
	})
	if err != nil {
		return fmt.Errorf("socket: calling emit: %w", err)
	}
	return nil
}

func (s *Socket) emitAck(id int, args ...any) error {
	err := s.adapter.Send(Packet{
		Type: "ack",
		Data: map[string]any{
			"id":   id,
			"args": ensureNonEmptyArgs(args),
		},
	})
	if err != nil {
		return fmt.Errorf("socket: calling emitAck: %w", err)
	}
	return nil
}

func (s *Socket) On(event string, fn func(args ...any)) *Socket {
	s.on(event, fn)
	return s
}

func (s *Socket) Off(event string, fn func(args ...any)) *Socket {
	s.off(event, fn)
	return s
}

func ensureNonEmptyArgs(args []any) []any {
	if len(args) == 0 {
		return []any{}
	}
	return args
}
